gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java

```diff
@@ -0,0 +1,166 @@
+  private OwnerAndPermission destinationOwnerAndPermission;
+  /**
+   * Desired {@link OwnerAndPermission} of the ancestor directories of the destination path. The list is ordered from
+   * deepest to highest directory.
+   *
+   * <p>
+   * For example, if {@link #destination} is /a/b/c/file, then the first element of this list is the desired owner and
+   * permission for directory /a/b/c, the second is the desired owner and permission for directory /a/b, and so on.
+   * </p>
+   *
+   * <p>
+   * If there are fewer elements in the list than ancestor directories in {@link #destination}, it is understood that
+   * extra directories are allowed to have any owner and permission.
+   * </p>
+   */
+  private List<OwnerAndPermission> ancestorsOwnerAndPermission;
+  /** Checksum of the origin file. */
+  private byte[] checksum;
+
+  @Override
+  public FileStatus getFileStatus() {
+    return this.origin;
+  }
+
+  @Override
+  public void write(DataOutput dataOutput) throws IOException {
+    this.origin.write(dataOutput);
+    Text.writeString(dataOutput, this.destination.toString());
+    Text.writeString(dataOutput, this.relativeDestination.toString());
+    this.destinationOwnerAndPermission.write(dataOutput);
+    dataOutput.writeInt(this.ancestorsOwnerAndPermission.size());
+    for (OwnerAndPermission oap : this.ancestorsOwnerAndPermission) {
+      oap.write(dataOutput);
+    }
+    dataOutput.writeInt(this.checksum.length);
+    dataOutput.write(this.checksum);
+  }
+
+  @Override
+  public void readFields(DataInput dataInput) throws IOException {
+    this.origin = new FileStatus();
+    this.origin.readFields(dataInput);
+    this.destination = new Path(Text.readString(dataInput));
+    this.relativeDestination = new Path(Text.readString(dataInput));
+    this.destinationOwnerAndPermission = OwnerAndPermission.read(dataInput);
+    int ancestors = dataInput.readInt();
+    this.ancestorsOwnerAndPermission = Lists.newArrayList();
+    for (int i = 0; i < ancestors; i++) {
+      this.ancestorsOwnerAndPermission.add(OwnerAndPermission.read(dataInput));
+    }
+    int checksumSize = dataInput.readInt();
+    this.checksum = new byte[checksumSize];
+    dataInput.readFully(this.checksum);
+  }
+
+  /**
+   * Read a {@link gobblin.data.management.copy.CopyableFile} from a {@link java.io.DataInput}.
+   *
+   * @throws IOException
+   */
+  public static CopyableFile read(DataInput dataInput) throws IOException {
+    CopyableFile copyableFile = new CopyableFile();
+    copyableFile.readFields(dataInput);
+    return copyableFile;
+  }
+
+  /**
+   * Serialize an instance of {@link CopyableFile} into a {@link String}. Usually use to store the copyableFile in state
+   * at key {@link #SERIALIZED_COPYABLE_FILE}
+   *
+   * @param copyableFile to be serialized
+   * @return serialized string
+   */
+  public static String serializeCopyableFile(CopyableFile copyableFile) throws IOException {
```

Comment:
Mainly that is more compact, and most Hadoop objects are already Writeable, which only supports byte serialization.
I am not opposed to json, but it will take more work I'm guessing.